File: sdk/worker/session.go
package worker

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"sync"
	"time"

	"git.saveweb.org/saveweb/hq/internal/trackerclient"
	"git.saveweb.org/saveweb/hq/pkg/protocol"
)

type Session struct {
	mu                      sync.Mutex
	config                  Config
	tracker                 *trackerclient.Client
	id                      string
	projectID               string
	leaseExpiresAt          int64
	agentHeartbeatSeconds   int64
	sessionHeartbeatSeconds int64
	assignment              *routeClient
	assignmentSessionLease  int64
	lastBackgroundError     error
	ctx                     context.Context
	cancel                  context.CancelFunc
	loops                   sync.WaitGroup
	claimsPaused            bool
	localAdmin              *LocalAdmin
	closed                  bool
}

func OpenSession(ctx context.Context, config Config, projectID string, attrs protocol.Attrs) (*Session, error) {
	config, err := config.normalized()
	if err != nil {
		return nil, err
	}
	if projectID == "" || attrs == nil {
		return nil, fmt.Errorf("worker: project ID and attrs are required")
	}
	attrs, err = cloneAttrs(attrs)
	if err != nil {
		return nil, err
	}
	tracker, err := trackerFor(config)
	if err != nil {
		return nil, err
	}
	agent, err := tracker.UpsertAgent(ctx, protocol.AgentUpsertRequest{
		Kind: protocol.AgentKindWorker, Name: config.AgentName, Version: config.AgentVersion,
		Attrs: attrs,
	})
	if err != nil {
		return nil, convertTrackerError(err)
	}
	session, err := tracker.CreateSession(ctx, protocol.CreateSessionRequest{ProjectID: projectID, Attrs: attrs})
	if err != nil {
		return nil, convertTrackerError(err)
	}
	loopContext, cancel := context.WithCancel(context.Background())
	result := &Session{
		config: config, tracker: tracker, id: session.SessionID, projectID: projectID,
		leaseExpiresAt:          session.LeaseExpiresAt,
		agentHeartbeatSeconds:   agent.HeartbeatAfterSeconds,
		sessionHeartbeatSeconds: session.HeartbeatAfterSeconds,
		ctx:                     loopContext, cancel: cancel,
	}
	result.loops.Add(2)
	go result.agentHeartbeatLoop(attrs)
	go result.sessionHeartbeatLoop()
	return result, nil
}

func (s *Session) ID() string { return s.id }

func (s *Session) ProjectID() string { return s.projectID }

// SubmitReceiver writes one immutable jobs-jsonl-zstd-v1 object through the
// trusted tracker gateway. A transport ambiguity may create a duplicate object;
// later stages deduplicate by stable job ID.
func (s *Session) SubmitReceiver(
	ctx context.Context,
	receiverID string,
	jobs []protocol.JobSpecV1,
) (protocol.ReceiverBatchResponse, error) {
	s.mu.Lock()
	if s.closed {
		s.mu.Unlock()
		return protocol.ReceiverBatchResponse{}, ErrSessionClosed
	}
	sessionID := s.id
	s.mu.Unlock()
	result, err := s.tracker.SubmitReceiverBatch(ctx, receiverID, protocol.ReceiverBatchRequest{
		SessionID: sessionID, Jobs: jobs,
	})
	if err != nil {
		return protocol.ReceiverBatchResponse{}, convertTrackerError(err)
	}
	if result.ProjectID != s.projectID || result.ReceiverID != receiverID ||
		result.Format != "jobs-jsonl-zstd-v1" || result.JobsCount != int64(len(jobs)) {
		return protocol.ReceiverBatchResponse{}, fmt.Errorf("worker: tracker returned a mismatched receiver result")
	}
	return result, nil
}

func (s *Session) Close() error {
	s.mu.Lock()
	if s.closed {
		s.mu.Unlock()
		return nil
	}
	s.closed = true
	assignment := s.assignment
	s.assignment = nil
	admin := s.localAdmin
	s.localAdmin = nil
	s.cancel()
	s.mu.Unlock()
	if admin != nil {
		_ = admin.Close()
	}
	s.loops.Wait()
	if assignment != nil {
		assignment.Close()
	}
	return nil
}

func (s *Session) SetClaimsPaused(value bool) {
	s.mu.Lock()
	s.claimsPaused = value
	s.mu.Unlock()
}

func (s *Session) ClaimsPaused() bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.claimsPaused
}

func (s *Session) LastBackgroundError() error {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.lastBackgroundError
}

func (s *Session) reportBackgroundError(err error) {
	err = convertTrackerError(err)
	s.mu.Lock()
	s.lastBackgroundError = err
	handler := s.config.OnBackgroundError
	s.mu.Unlock()
	if handler != nil {
		handler(err)
	}
}

func (s *Session) agentHeartbeatLoop(attrs protocol.Attrs) {
	defer s.loops.Done()
	interval := s.agentHeartbeatSeconds
	if interval < 1 {
		interval = 30
	}
	timer := time.NewTimer(time.Duration(interval) * time.Second)
	defer timer.Stop()
	for {
		select {
		case <-s.ctx.Done():
			return
		case <-timer.C:
		}
		ctx, cancel := context.WithTimeout(s.ctx, s.config.RequestTimeout)
		response, err := s.tracker.HeartbeatAgent(ctx, protocol.AgentHeartbeatRequest{
			Version: s.config.AgentVersion, Attrs: attrs,
		})
		cancel()
		next := int64(5)
		if err != nil {
			s.reportBackgroundError(err)
		} else {
			next = response.HeartbeatAfterSeconds
			if next < 1 {
				next = 30
			}
		}
		timer.Reset(time.Duration(next) * time.Second)
	}
}

func (s *Session) sessionHeartbeatLoop() {
	defer s.loops.Done()
	interval := s.sessionHeartbeatSeconds
	if interval < 1 {
		interval = 30
	}
	timer := time.NewTimer(time.Duration(interval) * time.Second)
	defer timer.Stop()
	for {
		select {
		case <-s.ctx.Done():
			return
		case <-timer.C:
		}
		ctx, cancel := context.WithTimeout(s.ctx, s.config.RequestTimeout)
		response, err := s.tracker.HeartbeatSession(ctx, s.id)
		cancel()
		next := int64(5)
		if err != nil {
			s.reportBackgroundError(err)
		} else {
			s.mu.Lock()
			s.leaseExpiresAt = response.LeaseExpiresAt
			s.mu.Unlock()
			next = response.HeartbeatAfterSeconds
			if next < 1 {
				next = 30
			}
		}
		timer.Reset(time.Duration(next) * time.Second)
	}
}

func (s *Session) route(ctx context.Context, acceptTypes []string, expected *routeIdentity, force bool) (*routeClient, error) {
	s.mu.Lock()
	if s.closed {
		s.mu.Unlock()
		return nil, ErrSessionClosed
	}
	now := time.Now().Unix()
	if !force && s.assignment != nil && s.assignment.assignment.AccessTokenExpires > now+30 &&
		s.assignmentSessionLease == s.leaseExpiresAt && (expected == nil || expected.matches(s.assignment.assignment)) {
		result := s.assignment
		s.mu.Unlock()
		return result, nil
	}
	leaseAtRequest := s.leaseExpiresAt
	s.mu.Unlock()

	response, err := s.tracker.GetAssignment(ctx, protocol.GetAssignmentRequest{
		SessionID: s.id, AcceptTypes: acceptTypes,
	})
	if err != nil {
		return nil, convertTrackerError(err)
	}
	if response.Assignment == nil {
		if expected != nil {
			return nil, ErrRouteRetired
		}
		return nil, ErrNoAssignment
	}
	if expected != nil && !expected.matches(*response.Assignment) {
		return nil, ErrRouteRetired
	}
	created, err := newRouteClient(*response.Assignment, s.config)
	if err != nil {
		return nil, err
	}
	s.mu.Lock()
	if s.closed {
		s.mu.Unlock()
		created.Close()
		return nil, ErrSessionClosed
	}
	previous := s.assignment
	s.assignment = created
	s.assignmentSessionLease = leaseAtRequest
	s.mu.Unlock()
	if previous != nil {
		previous.Close()
	}
	return created, nil
}

func cloneAttrs(value protocol.Attrs) (protocol.Attrs, error) {
	encoded, err := json.Marshal(value)
	if err != nil {
		return nil, fmt.Errorf("worker: attrs are not JSON encodable: %w", err)
	}
	var result protocol.Attrs
	decoder := json.NewDecoder(strings.NewReader(string(encoded)))
	decoder.UseNumber()
	if err := decoder.Decode(&result); err != nil {
		return nil, fmt.Errorf("worker: clone attrs: %w", err)
	}
	return result, nil
}

func convertTrackerError(err error) error {
	var trackerError *trackerclient.Error
	if errors.As(err, &trackerError) {
		return &APIError{Status: trackerError.Status, API: trackerError.API}
	}
	return err
}

File: sdk/python/src/saveweb_hq/session.py
"""Synchronous worker session and generation-bound queue batches."""

from __future__ import annotations

import copy
import threading
import time
from dataclasses import dataclass
from typing import Any, Callable

from .errors import (
    APIError,
    NoAssignmentError,
    RouteRetiredError,
    SessionClosedError,
    TransportError,
)
from .transport import JSONTransport, TrackerClient, _dumps, _loads


@dataclass(frozen=True, slots=True)
class Config:
    tracker_url: str
    machine_token: str
    agent_id: str
    agent_name: str = "Saveweb worker"
    agent_version: str = "python-sdk-dev"
    allow_http_tracker: bool = False
    allow_http_shard: bool = False
    request_timeout: float = 45.0
    on_background_error: Callable[[Exception], None] | None = None

    def validate(self) -> None:
        if not self.tracker_url or not self.machine_token or not self.agent_id:
            raise ValueError("tracker URL and machine credentials are required")
        if not 1.0 <= self.request_timeout <= 600.0:
            raise ValueError("request_timeout must be between 1 and 600 seconds")


@dataclass(frozen=True, slots=True)
class Route:
    project_id: str
    shard_id: str
    generation: int


@dataclass(frozen=True, slots=True)
class _RouteIdentity:
    project_id: str
    shard_id: str
    generation: int
    owner_agent_id: str

    def matches(self, assignment: dict[str, Any]) -> bool:
        return (
            self.project_id == assignment.get("project_id")
            and self.shard_id == assignment.get("shard_id")
            and self.generation == assignment.get("generation")
            and self.owner_agent_id == assignment.get("owner_agent_id")
        )


class _RouteClient:
    def __init__(self, assignment: dict[str, Any], config: Config) -> None:
        try:
            self.identity = _RouteIdentity(
                project_id=str(assignment["project_id"]),
                shard_id=str(assignment["shard_id"]),
                generation=int(assignment["generation"]),
                owner_agent_id=str(assignment["owner_agent_id"]),
            )
            self.endpoint_version = int(assignment["endpoint_version"])
            self.access_token = str(assignment["access_token"])
            self.access_token_expires_at = int(assignment["access_token_expires_at"])
            endpoint = str(assignment["endpoint"])
            pin_value = assignment.get("tls_spki_sha256")
            pin = None if pin_value is None else str(pin_value)
        except (KeyError, TypeError, ValueError) as error:
            raise TransportError(f"invalid tracker assignment: {error}") from error
        if self.identity.generation < 1 or not self.access_token:
            raise TransportError("invalid tracker assignment identity or token")
        self._transport = JSONTransport(
            endpoint,
            timeout=config.request_timeout,
            allow_http=config.allow_http_shard,
            spki_pin=pin,
            max_request_bytes=1 << 20,
            max_response_bytes=16 << 20,
        )

    def close(self) -> None:
        self._transport.close()

    def request(self, endpoint: str, payload: dict[str, Any]) -> dict[str, Any]:
        return self._transport.request_json(
            "POST",
            endpoint,
            payload,
            {"Authorization": f"Bearer {self.access_token}"},
        )


class Session:
    def __init__(
        self,
        config: Config,
        tracker: TrackerClient,
        session_id: str,
        project_id: str,
        attrs: dict[str, Any],
        lease_expires_at: int,
        agent_heartbeat_seconds: int,
        session_heartbeat_seconds: int,
    ) -> None:
        self.config = config
        self.id = session_id
        self.project_id = project_id
        self._tracker = tracker
        self._attrs = attrs
        self._lease_expires_at = lease_expires_at
        self._agent_heartbeat_seconds = max(agent_heartbeat_seconds, 1)
        self._session_heartbeat_seconds = max(session_heartbeat_seconds, 1)
        self._assignment: _RouteClient | None = None
        self._assignment_session_lease = 0
        self._last_background_error: Exception | None = None
        self._closed = False
        self._lock = threading.Lock()
        self._stop = threading.Event()
        self._threads = [
            threading.Thread(target=self._agent_heartbeat_loop, daemon=True),
            threading.Thread(target=self._session_heartbeat_loop, daemon=True),
        ]
        for thread in self._threads:
            thread.start()

    @property
    def last_background_error(self) -> Exception | None:
        with self._lock:
            return self._last_background_error

    def __enter__(self) -> Session:
        return self

    def __exit__(self, *_: object) -> None:
        self.close()

    def close(self) -> None:
        with self._lock:
            if self._closed:
                return
            self._closed = True
            route = self._assignment
            self._assignment = None
            self._stop.set()
        for thread in self._threads:
            thread.join(timeout=self.config.request_timeout + 1.0)
        if route is not None:
            route.close()
        self._tracker.close()

    def claim(
        self,
        max_jobs: int = 64,
        lease_seconds: int = 300,
        accept_types: list[str] | None = None,
    ) -> Batch:
        accepted = [] if accept_types is None else list(accept_types)
        route = self._route(accepted, expected=None, force=False)
        request = {
            **self._route_request(route.identity),
            "max_jobs": max_jobs,
            "lease_seconds": lease_seconds,
            "accept_types": accepted,
        }
        response = route.request("/api/v1/queue/claim", request)
        if (
            response.get("project_id") != route.identity.project_id
            or response.get("shard_id") != route.identity.shard_id
            or response.get("generation") != route.identity.generation
        ):
            raise TransportError("shard returned a mismatched claim route")
        jobs = response.get("jobs")
        if not isinstance(jobs, list):
            raise TransportError("claim response jobs must be a list")
        return Batch(self, route.identity, jobs)

    def _route(
        self,
        accept_types: list[str],
        *,
        expected: _RouteIdentity | None,
        force: bool,
    ) -> _RouteClient:
        with self._lock:
            if self._closed:
                raise SessionClosedError("worker session is closed")
            now = int(time.time())
            if (
                not force
                and self._assignment is not None
                and self._assignment.access_token_expires_at > now + 30
                and self._assignment_session_lease == self._lease_expires_at
                and (expected is None or expected == self._assignment.identity)
            ):
                return self._assignment
            lease_at_request = self._lease_expires_at
        response = self._tracker.get_assignment(self.id, accept_types)
        assignment = response.get("assignment")
        if assignment is None:
            if expected is not None:
                raise RouteRetiredError("claimed shard route is no longer assigned")
            raise NoAssignmentError("no active shard assignment")
        if not isinstance(assignment, dict):
            raise TransportError("tracker assignment must be an object or null")
        if expected is not None and not expected.matches(assignment):
            raise RouteRetiredError("claimed shard owner or generation changed")
        created = _RouteClient(assignment, self.config)
        with self._lock:
            if self._closed:
                created.close()
                raise SessionClosedError("worker session is closed")
            previous = self._assignment
            self._assignment = created
            self._assignment_session_lease = lease_at_request
        if previous is not None:
            previous.close()
        return created

    def _route_request(self, identity: _RouteIdentity) -> dict[str, Any]:
        return {
            "project_id": identity.project_id,
            "shard_id": identity.shard_id,
            "generation": identity.generation,
            "session_id": self.id,
        }

    def _report_background_error(self, error: Exception) -> None:
        with self._lock:
            self._last_background_error = error
            callback = self.config.on_background_error
        if callback is not None:
            try:
                callback(error)
            except Exception as callback_error:  # noqa: BLE001 - keep heartbeat thread alive.
                with self._lock:
                    self._last_background_error = callback_error

    def _agent_heartbeat_loop(self) -> None:
        interval = float(self._agent_heartbeat_seconds)
        while not self._stop.wait(interval):
            try:
                response = self._tracker.heartbeat_agent(
                    {"version": self.config.agent_version, "attrs": self._attrs}
                )
                interval = float(max(int(response["heartbeat_after_seconds"]), 1))
            except Exception as error:  # noqa: BLE001 - surfaced through the SDK callback.
                self._report_background_error(error)
                interval = 5.0

    def _session_heartbeat_loop(self) -> None:
        interval = float(self._session_heartbeat_seconds)
        while not self._stop.wait(interval):
            try:
                response = self._tracker.heartbeat_session(self.id)
                with self._lock:
                    self._lease_expires_at = int(response["lease_expires_at"])
                interval = float(max(int(response["heartbeat_after_seconds"]), 1))
            except Exception as error:  # noqa: BLE001 - surfaced through the SDK callback.
                self._report_background_error(error)
                interval = 5.0


class Batch:
    def __init__(self, session: Session, identity: _RouteIdentity, jobs: list[Any]) -> None:
        self._session = session
        self._identity = identity
        if not all(isinstance(job, dict) for job in jobs):
            raise TransportError("claim response contains a non-object job")
        self.jobs: tuple[dict[str, Any], ...] = tuple(copy.deepcopy(jobs))

    @property
    def route(self) -> Route:
        return Route(
            project_id=self._identity.project_id,
            shard_id=self._identity.shard_id,
            generation=self._identity.generation,
        )

    def refresh(self) -> None:
        self._session._route([], expected=self._identity, force=True)

    def complete(self, items: list[dict[str, Any]]) -> dict[str, Any]:
        return self._mutation("/api/v1/queue/complete", {"items": items})

    def fail(self, items: list[dict[str, Any]]) -> dict[str, Any]:
        return self._mutation("/api/v1/queue/fail", {"items": items})

    def extend_lease(self, extend_seconds: int, items: list[dict[str, Any]]) -> dict[str, Any]:
        return self._mutation(
            "/api/v1/queue/extend-lease",
            {"extend_seconds": extend_seconds, "items": items},
        )

    def _mutation(self, endpoint: str, fields: dict[str, Any]) -> dict[str, Any]:
        route = self._session._route([], expected=self._identity, force=False)
        payload = {**self._session._route_request(self._identity), **fields}
        try:
            return route.request(endpoint, payload)
        except (APIError, TransportError) as operation_error:
            if isinstance(operation_error, APIError) and operation_error.code not in {
                "stale_generation",
                "shard_not_active",
                "owner_lease_expired",
                "shard_unavailable",
                "invalid_access_token",
                "session_expired",
            }:
                raise
            try:
                self._session._route([], expected=self._identity, force=True)
            except RouteRetiredError:
                raise RouteRetiredError(
                    "claimed route was retired; discard this outcome"
                ) from operation_error
            except Exception:  # noqa: BLE001 - the original operation error is authoritative.
                pass
            raise


def open_session(
    config: Config,
    project_id: str,
    attrs: dict[str, Any] | None = None,
) -> Session:
    config.validate()
    if not project_id:
        raise ValueError("project_id is required")
    cloned_attrs = _loads(_dumps({} if attrs is None else attrs))
    if not isinstance(cloned_attrs, dict):
        raise ValueError("attrs must be an object")
    tracker = TrackerClient(
        config.tracker_url,
        config.machine_token,
        config.agent_id,
        timeout=config.request_timeout,
        allow_http=config.allow_http_tracker,
    )
    try:
        agent = tracker.upsert_agent(
            {
                "kind": "worker",
                "name": config.agent_name,
                "version": config.agent_version,
                "attrs": cloned_attrs,
                "endpoint": None,
                "endpoint_version": None,
                "tls_spki_sha256": None,
            }
        )
        session = tracker.create_session({"project_id": project_id, "attrs": cloned_attrs})
        try:
            agent_data = agent["agent"]
            if not isinstance(agent_data, dict) or agent_data.get("kind") != "worker":
                raise ValueError("agent is not a worker")
            return Session(
                config=config,
                tracker=tracker,
                session_id=str(session["session_id"]),
                project_id=project_id,
                attrs=cloned_attrs,
                lease_expires_at=int(session["lease_expires_at"]),
                agent_heartbeat_seconds=int(agent["heartbeat_after_seconds"]),
                session_heartbeat_seconds=int(session["heartbeat_after_seconds"]),
            )
        except (KeyError, TypeError, ValueError) as error:
            raise TransportError(f"tracker returned invalid session metadata: {error}") from error
    except Exception:
        tracker.close()
        raise
